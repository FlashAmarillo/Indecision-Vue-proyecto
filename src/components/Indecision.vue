<template>
  <img :src="image" alt="imagen-background" >
  <div class="bg-dark"></div>
  <div class="indecision-container">
      <input 
        type="text" 
        placeholder="Hazme una pregunta" 
        v-model="question" >
      <p>Recuerda terminar con un signo de interrogaciÃ³n (?)</p>

      <div v-if="isValidQuestion">
          <h2> {{ question }} </h2>
          <h1>{{ answer }}</h1>
      </div>
  </div>
</template>

<script>
export default {
    data() {
        return {
            question: null,
            answer: null,
            image: null,
            isValidQuestion: false,
        }
    },
    methods: {
        async getAnswer() {

            try {
                this.answer = 'Pensando...'
    
                //url de la api de yesno.wtf/
                const url = 'https://yesno.wtf/api'
                
                //hacemos la peticion a la API
                const { answer, image } = await fetch(url).then(respuesta => respuesta.json())
    
                //insertamos los valores en las propiedades de la data
                this.answer = answer === 'yes' ? 'Si!!ðŸ”¥' : 'No!'
                this.image = image

            } catch (error) {
                this.answer = 'No se pudo cargar del API';
                this.image = null;
                console.log(error)
            }
        }
    },
    watch: { //va a estar observando los valores que les indiques y ejecuta una funcion cada vez que cambie
        question( value, oldValue) {

            this.isValidQuestion = false;

            if( !value.includes('?')) return;

            this.isValidQuestion = true;

            // TODO: realizar peticion HTTP
            this.getAnswer();
        }
    }
}
</script>

<style scoped>

/* scoped esto hace que el siguiente CSS solo sea aplicado apra este componente */

    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
        margin-bottom: 1rem;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>