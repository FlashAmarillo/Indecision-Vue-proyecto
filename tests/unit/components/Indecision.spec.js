import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision.vue";

describe('Funcionalidad del componente Indesicion', () => {

    let wrapper;

    // definimos un console.log spy para estudiar si se dispara un console.log en nuestro componente
    let clgSpy;

    // Construyendo un Moch para fetch API ya que no existe la funcion en Nodejs
    global.fetch = jest.fn( () => Promise.resolve({
        json: () => Promise.resolve({
            answer: "yes",
            forced: false,
            image: "https://yesno.wtf/assets/yes/2.gif"
        })
    }));

    beforeEach( () => {
        wrapper = shallowMount( Indecision );
        
        clgSpy = jest.spyOn( console, 'log');

        jest.clearAllMocks();
    
    })

    test('Debe hacer match con el snapshot', () => { 
        expect( wrapper.html() ).toMatchSnapshot();
    })

    test('Escribir en el input NO debe dispara nada', async () => {

        const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer' )

        // hacemos la referencia al input
        const input = wrapper.find('input');

        // Le agregamos un valor al input
        await input.setValue('Hola Mundo');

        // llamo a mi spy para ver si ejecuto el console.log al menos 1 vez
        expect( clgSpy ).toHaveBeenCalledTimes(1);

        // compruebo que el getAnswer NO  se dispare
        expect(getAnswerSpy).toHaveBeenCalledTimes(0);
    })

    test('escribir el simbolo (?) debe disparar la funcion getAnswer', async () => {
        
        // definimos la funcion asincrona getAnswer 
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');

        //hacemos referencia al input
        const input = wrapper.find('input');

        //le agregamos un simbolo de ? 
        await input.setValue('llovera maÃ±ana ?');

        // la prueba funciona correctamente al hacer mocking de la funcion fetch API
        expect(getAnswerSpy).toHaveBeenCalled();
    })

    test('pruebas en getAnswer', async () => {
        
        await wrapper.vm.getAnswer();

        const img = wrapper.find('img');

        expect( img.exists() ).toBeTruthy();
        expect( wrapper.vm.image ).toBe('https://yesno.wtf/assets/yes/2.gif');
        expect( wrapper.vm.answer ).toBe('Si!!ðŸ”¥');
    })

    test('pruebas en getAnswer - fallos en el API', async () => {
        
        fetch.mockImplementationOnce( () => Promise.reject('API is down') );
        
        await wrapper.vm.getAnswer();

        const img = wrapper.find('image');

        // Si todo sale mal
        expect( img.exists() ).toBeFalsy();
        expect( wrapper.vm.answer ).toBe('No se pudo cargar del API');
    })
})