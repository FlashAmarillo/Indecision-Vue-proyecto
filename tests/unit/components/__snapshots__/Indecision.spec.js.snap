// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Funcionalidad del componente Indesicion Debe hacer match con el snapshot 1`] = `
<img alt="imagen-background">
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>
`;
